src/utils.c

```diff
@@ -743,3 +743,169 @@ gboolean r_fakeroot_cleanup(const gchar *envpath, GError **error)
+			if (*build == NULL)
+				goto out;
+		}
+		*build_pos = '\0';
+	}
+
+	pre_release_pos = g_strrstr(version_copy, "-");
+	if (pre_release_pos != NULL) {
+		if (pre_release != NULL) {
+			*pre_release = g_strdup(pre_release_pos + 1);
+			if (*pre_release == NULL)
+				goto out;
+		}
+		*pre_release_pos = '\0';
+	}
+
+	version_parts = g_strsplit(version_copy, ".", 3);
+	// Note that g_strsplit returns an empty array when splitting: ""
+	if (version_parts == NULL)
+		goto out;
+	// and NULL terminates arrays it creates
+	if (version_parts[0] == NULL)
+		goto out;
+	// only on component -> shift right
+	if (version_parts[1] == NULL) {
+		version_parts[2] = version_parts[0];
+		version_parts[0] = NULL;
+	}
+	// only two components -> shift right
+	if (version_parts[1] != NULL && version_parts[2] == NULL) {
+		version_parts[2] = version_parts[1];
+		version_parts[1] = version_parts[0];
+		version_parts[0] = NULL;
+	}
+
+	for (int i = 0; i < 3; i++) {
+		if (version_parts[i] != NULL) {
+			// Note that negative numbers are already implicitly excluded due to the search for the '-pre_release'
+			version_core[i] = g_ascii_strtoll(version_parts[i], NULL, 10);
+			// handle g_ascii_strtoll returning zero on a conversion error
+			if ((version_core[i] == 0) && (g_strcmp0(version_parts[i], "0") != 0)) {
+				g_strfreev(version_parts);
+				goto out;
+			}
+		} else
+			version_core[i] = 0;
+	}
+
+	g_strfreev(version_parts);
+	return TRUE;
+
+out:
+	if (pre_release && *pre_release)
+		g_free(*pre_release);
+	if (build && *build)
+		g_free(*build);
+	return FALSE;
+}
+
+gboolean semver_less_equal(const guint version_a[3], gchar *pre_release_a, const guint version_b[3], gchar *pre_release_b)
+{
+	gchar **pre_fields_a = NULL;
+	gchar **pre_fields_b = NULL;
+	gboolean is_num_a = FALSE;
+	gboolean is_num_b = FALSE;
+	guint num_a = 0;
+	guint num_b = 0;
+	gint cmp = 0;
+	gboolean result = FALSE;
+	int i = 0;
+
+	// compare version cores: major, minor, patch
+	for (i = 0; i < 3; i++) {
+		if (version_a[i] < version_b[i]) {
+			return TRUE;
+		} else if (version_a[i] > version_b[i]) {
+			return FALSE;
+		}
+	}
+
+	// version cores are equal, compare pre-release identifiers
+	if (pre_release_a == NULL && pre_release_b == NULL) {
+		return TRUE;
+	} else if (pre_release_a == NULL) {
+		return FALSE; // version_a > version_b-pre_release
+	} else if (pre_release_b == NULL) {
+		return TRUE; // version_a-pre_release < version_b
+	}
+
+	// compare	dot-separated fields of pre-release identifiers
+	pre_fields_a = g_strsplit(pre_release_a, ".", 0);
+	pre_fields_b = g_strsplit(pre_release_b, ".", 0);
+
+	i = 0;
+	while (pre_fields_a[i] != NULL && pre_fields_b[i] != NULL) {
+		is_num_a = g_ascii_isdigit(pre_fields_a[i][0]);
+		is_num_b = g_ascii_isdigit(pre_fields_b[i][0]);
+
+		if (is_num_a && is_num_b) {
+			//	compare numerically
+			num_a = g_ascii_strtoull(pre_fields_a[i], NULL, 10);
+			num_b = g_ascii_strtoull(pre_fields_b[i], NULL, 10);
```

Comment:
This can overflow on 32-bit architectures. All the core versions should probably be `guint64` to match the return value of `g_ascii_strtoull`.

Reply:
fixed in v3